### PR TITLE
Public output for ZkProgram

### LINKED ISSUE #523: ZkProgram methods return their return values rather than require them as arguments
Specifically, we'd want methods like this recursive add rollup method:

To go from:

```typescript
method(
        publicInput: Field,
        left: SelfProof<Field>,
        right: SelfProof<Field>
      ) {
        left.verify();
        right.verify();
        // assert that the left and right equal this input
        left.publicInput.add(right.publicInput).assertEquals(publicInput);
      },
```

to

```typescript
method(
        left: SelfProof<Field>,
        right: SelfProof<Field>
      ): Field  {
        left.verify();
        right.verify();

        return left.publicInput.add(right.publicInput);
      },
```

This would make the API more "snarky"

### PR DESCRIPTION
**Pull request by mitschabaude**
closes #523
bindings companion: https://github.com/o1-labs/snarkyjs-bindings/pull/4

* adapt the Pickles interface in `snarky.d.ts` to take both public input + output, instead of just input, to follow the changes in https://github.com/o1-labs/snarkyjs-bindings/pull/4
* propagate these changes to the `Proof` class, which is now `Proof<PublicInput, PublicOutput>`
* also propagate the changes to the `ZkProgram` API
* do not change `SmartContract`, which has the public output fixed to `null`

testing:
* [x] check that updating the Pickles spec from `Input` to `Input_and_output` with an empty output type doesn't break verification of zkapp proofs on the current network https://berkeley.minaexplorer.com/transaction/5Jv4brtooog2XxCoYqXwYDBqF39gzjQXKAc3t1234sQC5TEQQGDx

docs:
* [x] changelog (in https://github.com/o1-labs/snarkyjs/pull/876)
* [ ] update recursion docs

### CODE REVIEW

File: src/lib/proof_system.ts [hunk 1]
```diff
@@ -494,24 +599,39 @@ function picklesRuleFromFunction(
           : emptyWitness(type);
       } else if (arg.type === 'proof') {
         let Proof = proofArgs[arg.index];
-        let publicInput = getPublicInputType(Proof).fromFields(
-          previousInputs[arg.index]
-        );
-        let proofInstance: Proof<any>;
+        // split in input & output
+        // TODO: would be nice to either make Pickles return the right split or completely move
```
Comment (mitschabaude) [inline src/lib/proof_system.ts:603]: this TODO is addressed in a follow-up PR: https://github.com/o1-labs/snarkyjs/pull/876

File: src/lib/proof_system.ts [hunk 2]
```diff
@@ -494,24 +599,39 @@ function picklesRuleFromFunction(
...
+        // getting the public input & outputs to the JS side, from `snarkContext`, like private inputs
```
Comment (mitschabaude) [inline src/lib/proof_system.ts:604]: this TODO is addressed in a follow-up PR: https://github.com/o1-labs/snarkyjs/pull/876

**Review by Trivo25 — APPROVED** [review]
this is great!

File: src/examples/zkapps/simple_zkapp_with_proof.ts [hunk 3]
```diff
@@ -32,7 +32,7 @@ class NotSoSimpleZkapp extends SmartContract {
 
   @method update(
     y: Field,
-    oldProof: SelfProof<ZkappPublicInput>,
+    oldProof: SelfProof<ZkappPublicInput, null>,
```
[the remark this answers is not in the document]
Comment (mitschabaude) [reply src/examples/zkapps/simple_zkapp_with_proof.ts:35]: did this in the follow up PR: https://github.com/o1-labs/snarkyjs/pull/876/commits/8e4527e76f7ee55645a2770dba27218805216eb4

**Pull request merged**